### ISSUE TITLE
Issue with reading for some words

### ISSUE THREAD

**Issue by oldmerkum** [jrfonseca/jmdict-kindle#32]
Lookup for some words don't provide the common reading e.g.
人間の住める環境とはいえない場所に...
人間 lookup is じんかん instead of にんげん

**Comment by jrfonseca** [comment]
This is JMdict's entry for "にんげん":

```xml
<entry>
<ent_seq>1366770</ent_seq>
<k_ele>
<keb>人間</keb>
<ke_pri>ichi1</ke_pri>
<ke_pri>news1</ke_pri>
<ke_pri>nf01</ke_pri>
</k_ele>
<r_ele>
<reb>にんげん</reb>
<re_pri>ichi1</re_pri>
<re_pri>news1</re_pri>
<re_pri>nf01</re_pri>
</r_ele>
<sense>
<pos>&n;</pos>
<gloss>human being</gloss>
<gloss>human</gloss>
<gloss>person</gloss>
<gloss>man</gloss>
<gloss>mankind</gloss>
<gloss>humankind</gloss>
</sense>
<sense>
<pos>&n;</pos>
<gloss>character (of a person)</gloss>
</sense>
</entry>
```

And this is "じんかん":

```xml
<entry>
<ent_seq>2065280</ent_seq>
<k_ele>
<keb>人間</keb>
</k_ele>
<r_ele>
<reb>じんかん</reb>
</r_ele>
<sense>
<pos>&n;</pos>
<gloss>the world</gloss>
</sense>
</entry>
```

http://ftp.usf.edu/pub/ftp.monash.edu.au/pub/nihongo/edict_doc.html explains that the `ke_pri` and `re_pri` have priority information.

the jmdict.py script  looks at `re_pri`, but we seem to ignore `ke_pri`.  But the problem looks deeper than  that.

I don't quite understand how these incongruences are getting resolved, if at all.

 I'm afraid I haven't look at this code for so long I forgot all about it...  I doubt I'll have the time to dive into this soon. 
  Hopefully a charitable sole will help here...

**Comment by jrfonseca** [comment]
For reference this was a local debugging change I applied to get `jmdict.py` script to ignore anything but the word in question:

```diff
diff --git a/jmdict.py b/jmdict.py
index f431d17..e9972e5 100755
--- a/jmdict.py
+++ b/jmdict.py
@@ -231,7 +231,9 @@ class JMdictParser(XmlParser):
         self.element_start("JMdict")
         while self.token.type == XML_ELEMENT_START:
             entry = self.parse_entry()
-            entries.append(entry)
+            for kanji in entry.kanjis:
+                if kanji.keb == '人間':
+                    entries.append(entry)
             if len(entries) >= MAX_ENTRIES:
                 return entries
         self.element_end("JMdict")
```

**Comment by jrfonseca** [comment]
It looks like `jmdict.py` actually emits duplicate `<idx:orth value=人間 />` index entries in the Kindle dictionary.  So which one is used is a matter of chance.

I think `jmdict.py` is either missing logic to pick the highest priority sense, or the logic exists but is broken somehow...

**Comment by oldmerkum** [comment]
Looks like this function is where to start looking in `jmdict.py`:
```python
class JMdictParser(XmlParser):
...
    def parse_rank(self):
        re_pri = self.element_character_data("re_pri")
        if re_pri in ("news1", "ichi1", "spec1", "gai1"):
            return 0
        else:
            return 1
```

If you don't have time I can look more into it tonight.

**Comment by jrfonseca** [comment]
`dictionary.py`'s `sort_function` function is sorting the kana readings so the highest ranking (rank=0) are listed first.

But this doesn't help on this particular case, where the same kanji has two completely different kana readings with different meanings/ranking.

Furthermore, because the entries are split according to the first kana character, the first one will be the one with that starts with the kana character that happens be alphabetically first.


I think the solution is will imply adding a pass through the dictionary entries, that finds separate entries with colliding kanji readings, and prune kanji reading of the lowest priority, or merge the low-priority entry into the high priority entry as new sense?

> If you don't have time I can look more into it tonight.

Thank you.

**Comment by oldmerkum** [comment]
Just want to make sure I understand. So, my kindle doesn't seem to show more than one definition for a word, it finds a match and then shows it.
You're saying that one solution is to run through the dictionary and look for words which have identical kanji, and merge their meanings/readings into a single dictionary entry? (Ideally with the more common one first).
Otherwise, the lookup is at the whim of whichever entry comes up first (again, because only one entry result is returned when using kindle dictionary)

**Comment by jrfonseca** [comment]
That's right.

With my patch (a temporary hack to help debug this issue) from https://github.com/jmdict-kindle/jmdict-kindle/issues/32#issuecomment-2005130795 , jmdict.py products two files: `entry-jmdict-じ.html`, and `entry-jmdict-に.html`.

This is `entry-jmdict-じ.html` (with some extra newlines to make it more readable):
```html
<!DOCTYPE html><html><head><meta http-equiv=Content-Type content="text/html; charset=utf-8"><link rel=stylesheet type=text/css href=style.css></head>
<body topmargin=0 bottommargin=0 leftmargin=0 rightmargin=0><mbp:frameset>
<idx:entry name=v scriptable=yes><p class=lab><span class=l>じ</span>
<span class=h>んかん</span>
 [0]【人間】</p>
<ul><li> <span class=pos>n</span>
 the world </li>
</ul>
<div class=ex><span class=exh>Examples:</span>
<div class=sen><span>どんなに財産があっても、貪欲な人間は満足できない。世界中の財宝を集めても幸福にはなれない。</span>
<br><span>No amount of wealth can satisfy a covetous man. All the treasure in the world won't make him happy.</span>
</div>
<div class=sen><span>犬の聴覚の敏感さを人間と比べると、はるかに上だ。</span>
<br><span>The hearing acuity of dogs is superior by far to that of humans.</span>
</div>
<div class=sen><span>多くの人間にとって三角関数を学ぶことは無意味であるように思える。</span>
<br><span>I think that for many people learning the trigonometric functions is meaningless.</span>
</div>
<div class=sen><span>それは、人間が哺乳動物であることに由来しています。</span>
<br><span>That is because human beings are mammals.</span>
</div>
<div class=sen><span>ふふっ・・・その子、人間じゃないの。ロボットなのよ。ア・ン・ド・ロ・イ・ド。</span>
<br><span>Ha-ha ... She isn't human. She's a robot; A-n-d-r-o-i-d.</span>
</div>
</div>
<idx:orth value=じんかん /><idx:orth value=人間 /></idx:entry>
<hr><mbp:pagebreak/></mbp:frameset>
</body>
</html>
```

And this is `entry-jmdict-に.html`:
```html
<!DOCTYPE html><html><head><meta http-equiv=Content-Type content="text/html; charset=utf-8"><link rel=stylesheet type=text/css href=style.css></head>
<body topmargin=0 bottommargin=0 leftmargin=0 rightmargin=0><mbp:frameset>
<idx:entry name=v scriptable=yes><p class=lab><span class=l>に</span>
<span class=h>んげ<span class=nas>&#176;</span>
ん</span>
 [0]【人間】</p>
<ul><li> <span class=pos>n</span>
 human being; human; person; man; mankind; humankind </li>
<li> <span class=pos>n</span>
 character (of a person) </li>
</ul>
<div class=ex><span class=exh>Examples:</span>
<div class=sen><span>どんなに財産があっても、貪欲な人間は満足できない。世界中の財宝を集めても幸福にはなれない。</span>
<br><span>No amount of wealth can satisfy a covetous man. All the treasure in the world won't make him happy.</span>
</div>
<div class=sen><span>犬の聴覚の敏感さを人間と比べると、はるかに上だ。</span>
<br><span>The hearing acuity of dogs is superior by far to that of humans.</span>
</div>
<div class=sen><span>多くの人間にとって三角関数を学ぶことは無意味であるように思える。</span>
<br><span>I think that for many people learning the trigonometric functions is meaningless.</span>
</div>
<div class=sen><span>それは、人間が哺乳動物であることに由来しています。</span>
<br><span>That is because human beings are mammals.</span>
</div>
<div class=sen><span>ふふっ・・・その子、人間じゃないの。ロボットなのよ。ア・ン・ド・ロ・イ・ド。</span>
<br><span>Ha-ha ... She isn't human. She's a robot; A-n-d-r-o-i-d.</span>
</div>
</div>
<idx:orth value=にんげん /><idx:orth value=人間 /></idx:entry>
<hr><mbp:pagebreak/></mbp:frameset>
</body>
</html>
```

`<idx:orth value=.../>`  are the extra tags that Kindle uses to do lookups.  The kana `idx:orth` are distinct for both entries.  But  you see that `<idx:orth value=人間 />` appears twice.  That is the bug.  Kindle will pick one and ignore the other, and it looks like it picks the one from ``entry-jmdict-じ.html` ( not sure exactly.)

And one solution is indeed to ensure that for every kanji reading with duplicate entries, only one `<idx:orth value=... />` is outputed -- the one with highest priority.

In short, in  `entry-jmdict-に.html` there should be no `<idx:orth value=人間 />`, but otherwise everything else should be left the same.

**Comment by oldmerkum** [comment]
Created a pull request and did some small testing on my kindle. Works okay so far.
#33
